File: permute_new.py
import copy
from pprint import pprint

"""
if string has length 1, return string
for range from 0 to length,
    insert the last character into the list of all possibilities

"""


def permute(list1, length):
    if length==1:
        return([list1])
    else:
        last_char = list1[length-1]
        new_length = length-1
        orig = permute(list1[0:length-1], new_length)
        final_list = []
        list_length = len(orig)
        for x in range(0,len(orig)):
            for i in range(0, length):
                base = copy.copy(orig[x])
                base.insert(i, last_char)
                final_list.append(base)
        return final_list
    
abc = 'abcd'
abc_list = list(abc)
res = permute(abc_list, 4)

pprint(res)
pprint(len(res))

# print(abc_list[0:0])

File: permute_new.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<string> permute(string str) {
  int len = str.length();
  vector<string> res;
  res.push_back(str);
  if (len==1) {
    return res;
  }
  else {
    string first_part = str.substr(0, len-1);
    vector<string> base = permute(first_part);
    int list_size = base.size();
    string last_str = str.substr(len-1, 1);
    vector<string> res;
    string temp = "";
    
    // a
    // ab -> ab, ba
    
    for(int j=0; j<list_size; j++) {
      for(int i=0; i<len; i++) {
        temp = base[j];
        temp.insert(i, last_str);
        res.push_back(temp);
      }
    }
    return res;
  }
}

int main(int argc, char* argv[]) {
  string s = "abc";
  vector<string> res = permute(s);
  for(vector<string>::iterator it=res.begin(); it!=res.end(); ++it) {
    cout << *it << endl;
  }
  cout << res.size() << endl;
  
}